File: Render/Framebuffer.cpp
#include "Framebuffer.hpp"

#include <cstring>

namespace Render {
    Framebuffer::Framebuffer(unsigned int width, unsigned int height)
        : mWidth(width), mHeight(height)
    {
        mBits.resize(mWidth * mHeight * 3);
        if(mBits.size() > 0) {
            std::memset(&mBits[0], 0, mBits.size());
        }
    }

    unsigned int Framebuffer::width() const
    {
        return mWidth;
    }

    unsigned int Framebuffer::height() const
    {
        return mHeight;
    }

    const unsigned char *Framebuffer::bits() const
    {
        return &mBits[0];
    }

    void Framebuffer::setPixel(unsigned int x, unsigned int y, const Object::Color &color)
    {
        mBits[(mWidth * y + x) * 3 + 0] = static_cast<unsigned char>(color.red() * 0xff);
        mBits[(mWidth * y + x) * 3 + 1] = static_cast<unsigned char>(color.green() * 0xff);
        mBits[(mWidth * y + x) * 3 + 2] = static_cast<unsigned char>(color.blue() * 0xff);
    }

    Object::Color Framebuffer::toneMap(const Object::Radiance &rad)
    {
        float red = rad.red() / (rad.red() + 1);
        float green = rad.green() / (rad.green() + 1);
        float blue = rad.blue() / (rad.blue() + 1);

        return Object::Color(red, green, blue);
    }
}

File: src/render/framebuffer.rs
use crate::object;

use object::Color;
use object::Radiance;

pub struct Framebuffer {
    pub width : usize,
    pub height : usize,
    pub bits : Vec<u8>
}

impl Framebuffer {
    pub fn new(width : usize, height : usize) -> Framebuffer {
        let mut bits = Vec::new();
        bits.resize(width * height, 0);

        return Framebuffer {width, height, bits};
    }

    pub fn set_pixel(&mut self, x : usize, y : usize, color : Color) {
        self.bits[(y*self.width + x) * 3 + 0] = (color.red * 255.0) as u8;
        self.bits[(y*self.width + x) * 3 + 1] = (color.green * 255.0) as u8;
        self.bits[(y*self.width + x) * 3 + 2] = (color.blue * 255.0) as u8;
    }

    pub fn tone_map(rad : Radiance) -> Color {
        let red = rad.red / (rad.red + 1.0);
        let green = rad.green / (rad.green + 1.0);
        let blue = rad.blue / (rad.blue + 1.0);

        return Color::new(red, green, blue);
    }
}